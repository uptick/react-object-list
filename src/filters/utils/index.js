export FilterComparison from './FilterComparison'
export FilterLabel from './FilterLabel'
export RemoveFilter from './RemoveFilter'
export makeFilter from './makeFilter'
export * from './utils'
