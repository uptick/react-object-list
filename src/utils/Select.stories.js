import React from 'react'
import { storiesOf } from '@storybook/react'
import { withInfo } from '@storybook/addon-info'

import Select from './Select'

storiesOf('object-list/Utils/Select', module)
  .addDecorator((story, context) => withInfo(
    'Wrapper for react-select to ensure consistency'
  )(story)(context))
  .add('has custom menu style', () => (
    <Select menuStyle={{border: '5px dashed blue'}} />
  ))
  .add('has custom menu container style', () => (
    <Select menuContainerStyle={{border: '5px solid red'}} />
  ))
