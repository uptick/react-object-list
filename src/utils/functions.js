/**
 * Filters columns by those in extraColumns to get the visible or optional fields
 * @param  {object[]} columns      Array of columns or arrays of columns
 * @param  {string[]} extraColumns Array of keys of columns currently visible
 * @param  {bool} optional         The value optional is set to in the particular column
 * @return {object[]}              Returns only those columns which are visible
 */
const getVisibleColumns = (columns, extraColumns = [], optional = false) => {
  return setColumnLabels(columns).reduce((acc, column) => {
    if (column.hasOwnProperty('columns')) {
      // check the next node down
      const visible = column.optional !== undefined && _checkVisibilityHelper(column, extraColumns, optional)
      if (column.optional === undefined || visible) {
        const columns = getVisibleColumns(column.columns, extraColumns, optional)
        if (columns.length || visible) {
          acc.push({...column, columns})
        }
      }
    } else if (Array.isArray(column)) {
      // is a grouped column, check each of the items in the group
      const columns = column.filter(leafColumn => _checkVisibilityHelper(leafColumn, extraColumns, optional))
      if (columns.length) {
        acc.push(columns)
      }
    } else if (_checkVisibilityHelper(column, extraColumns, optional)) {
      // is a leaf node, check individual column
      acc.push(column)
    }
    return acc
  }, [])
}

const _checkVisibilityHelper = (column, extraColumns, optional) => {
  return !!column.optional === optional ||
  extraColumns.includes(column.fieldKey)
}

/** get all the leaves of an unbalanced tree */
const getLeafColumns = (columns) => {
  return columns.reduce((acc, column) => {
    if (column.hasOwnProperty('columns')) {
      return acc.concat(getLeafColumns(column.columns))
    }
    acc.push(column)
    return acc
  }, [])
}

/**
 * Annotates each column with a _colSpan and _rowSpan in the tree
 * which is used when rendering them
 *
 * @param {object[]} columns  Array of columns
 */
const annotateSpans = columns => {
  const depth = _getTreeDepth(columns)
  _annotateColumnSpan(columns, depth)
}

/** depth first search annotating each column with a _colSpan or _rowSpan as appropriate */
const _annotateColumnSpan = (columns, depth) => {
  let sum = 0
  columns.forEach(column => {
    if (!column.hasOwnProperty('columns')) {
      column._colSpan = 1
      column._rowSpan = depth
    } else {
      column._colSpan = _annotateColumnSpan(column.columns, depth - 1)
    }
    sum += column._colSpan
  })
  return sum
}

/** breadth first search for tree depth, as it might be imbalanced */
const _getTreeDepth = (columns) => {
  if (!columns.length) return 0

  let newColumns = []
  columns.forEach(column => {
    if (column.hasOwnProperty('columns')) {
      newColumns = newColumns.concat(column.columns)
    }
  })
  return _getTreeDepth(newColumns) + 1
}

const setColumnLabels = (columns) => {
  return columns.map(column => {
    if (column.hasOwnProperty('columns')) {
      return {
        ...column,
        columns: setColumnLabels(column.columns),
        fieldKey: column.fieldKey || column.header.replace(' ', '').toLowerCase(),
      }
    } else if (Array.isArray(column)) {
      const labels = column.map(({dataKey, fieldKey}) => fieldKey || dataKey.substring(dataKey.lastIndexOf('.') + 1))
      const fieldKeys = column.map(({dataKey, fieldKey}) => fieldKey || dataKey.substring(dataKey.lastIndexOf('.') + 1))
      return column.map(column => {
        const label = column.label || labels.join('_').replace(' ', '').toLowerCase()
        const fieldKey = column.fieldKey || fieldKeys.join('_').replace(' ', '').toLowerCase()
        const labelledColumn = {...column, label, fieldKey}
        labels.push(labels.shift())
        fieldKeys.push(fieldKeys.shift())
        return labelledColumn
      })
    } else {
      return {
        ...column,
        label: (column.fieldKey || column.dataKey.substring(column.dataKey.lastIndexOf('.') + 1)).replace(' ', '').toLowerCase(),
        fieldKey: (column.fieldKey || column.dataKey.substring(column.dataKey.lastIndexOf('.') + 1)),
      }
    }
  })
}

const sortByName = (a, b) => {
  const textA = a.name.toLowerCase()
  const textB = b.name.toLowerCase()
  return ((textA < textB) ? -1 : (textA > textB) ? 1 : 0)
}

export {
  _getTreeDepth,
  getLeafColumns,
  setColumnLabels,
  annotateSpans,
  getVisibleColumns,
  sortByName,
}
