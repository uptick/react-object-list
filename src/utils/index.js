export loadingSpinner from './loading'
export ErrorMessage from './ErrorMessage'

export {
  DATE_FORMAT,
  DATETIME_FORMAT,
  SHORTDATE_FORMAT,
  SHORTDATETIME_FORMAT,
  API_DATE_FORMAT,
  allowedLookupExpressions,
} from './constants'
export {
  COLUMN_TYPE,
  COLUMN_BASE_TYPE,
  FILTER_BASE_TYPE,
  META_TYPE,
  STATUS_TYPE,
  STATUS_CHOICES,
  ALL_SELECTED,
  SELECTION_TYPE,
} from './proptypes'
export {
  getVisibleColumns,
  setColumnLabels,
  sortByName,
  getLeafColumns,
} from './functions'
