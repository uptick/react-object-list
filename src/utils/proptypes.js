import PropTypes from 'prop-types'

const STATUS_CHOICES = {
  loading: 'loading',
  error: 'error',
  done: 'done',
}
const STATUS_TYPE = PropTypes.oneOf(Object.values(STATUS_CHOICES))

const ALL_SELECTED = 'all'
function selectionTypeValidator(isRequired) {
  return function(props, propName, componentName) {
    const value = props[propName]
    if (isRequired && (value === null || value === undefined)) {
      return new Error(`Missing prop \`${propName}\`
      supplied to \`${componentName}\` expected one of ['${ALL_SELECTED}', obj]`)
    }
    let validObject = true
    if (typeof value === 'object') {
      validObject = Object.entries(value).every(
        ([key, value]) => (typeof key === 'number' || 'string') && typeof value === 'boolean')
    } else {
      validObject = false
    }

    if (value !== ALL_SELECTED && !validObject) {
      return new Error(`Invalid prop \`${propName}\` of \`${value}\`
        supplied to \`${componentName}\` expected one of ['${ALL_SELECTED}', obj]`)
    }
  }
}
const SELECTION_TYPE = selectionTypeValidator(false)
SELECTION_TYPE.isRequired = selectionTypeValidator(true)

const COLUMN_BASE_TYPE = {
  fieldKey: PropTypes.string, // used to uniquely identify this column
  dataKey: PropTypes.string, // used to access the data inside the data object
  header: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
  item: PropTypes.func,
  sortKey: PropTypes.string,
  optional: PropTypes.bool,
  /* string or function of a data row returning a string to add to row classes */
  rowClass: PropTypes.oneOfType([PropTypes.string, PropTypes.func]),
}

const COLUMN_TYPE = PropTypes.shape(COLUMN_BASE_TYPE)

const FILTER_BASE_TYPE = {
  /** Name of filter, used to render filter */
  name: PropTypes.string,
  /** Identifier key for filter */
  filterKey: PropTypes.string,
  /** If true, indicates the filter cannot be removed */
  permanent: PropTypes.bool,
}

const META_TYPE = PropTypes.shape({
  /** the current sortkeys applied using {key: asc_desc_bool } eg. {name: true} for ascending */
  sortKeys: PropTypes.arrayOf(PropTypes.object),
  /** current page highlighted on the pagination */
  currentPage: PropTypes.number,
  /** the maximum number of items shown on a page */
  perPage: PropTypes.number,
  /** the keys of the optional columns that are currently being displayed */
  extraColumns: PropTypes.arrayOf(PropTypes.string),
  /** the total number of items in the dataset if only displaying a subset */
  totalCount: PropTypes.number,
  /** the list of items currently selected */
  selection: PropTypes.array,
})

export {
  COLUMN_TYPE,
  COLUMN_BASE_TYPE,
  FILTER_BASE_TYPE,
  META_TYPE,
  STATUS_TYPE,
  STATUS_CHOICES,
  SELECTION_TYPE,
  ALL_SELECTED,
}
