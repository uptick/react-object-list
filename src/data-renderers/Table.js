import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ClassNames from 'classnames'

import TableHeader from './TableHeader'
import Overlay from './Overlay'
import AllSelector from '../types/AllSelector'
import Selector from '../types/Selector'
import { getVisibleColumns, annotateSpans, getLeafColumns } from '../utils/functions'
import { getValueFromAccessor, handleRowClick } from './utils'
import { STATUS_TYPE, STATUS_CHOICES, SELECTION_TYPE, ALL_SELECTED } from '../utils/proptypes'

export default class TableRenderer extends Component {
  static propTypes = {
    columns: PropTypes.array,
    meta: PropTypes.object,
    saveColumnWidth: PropTypes.func,
    data: PropTypes.array,
    summaryData: PropTypes.object,
    /** eg. { label: {width: "100"} } where 'label' is the label generated for the column */
    columnWidths: PropTypes.object,
    /** set the sort to the given sortkeys */
    updateSorting: PropTypes.func,
    /** IDs of currently selected items set to true if selected */
    selection: SELECTION_TYPE,
    /** Function to select one ID or an array of IDs */
    select: PropTypes.func,
    /** loading status used if data is loaded asynchronously  */
    status: STATUS_TYPE,
    /** Count of selected items */
    numSelected: PropTypes.number,
    /** Function called when table row is clicked */
    itemOnClick: PropTypes.func,
  }

  static defaultProps = {
    meta: {
      extraColumns: [],
      sortKeys: [],
    },
    status: STATUS_CHOICES.done,
    columnWidths: {},
    selection: {},
  }

  state = {
    columns: getVisibleColumns(this.props.columns, this.props.meta.extraColumns),
  }

  componentWillReceiveProps(nextProps) {
    // TODO: optimise performance and only do this if the columns prop has changed
    this.setState(() => ({
      columns: getVisibleColumns(nextProps.columns, nextProps.meta.extraColumns),
    }))
  }

  renderHeader = () => {
    const {columns} = this.state
    const {select, numSelected, data} = this.props

    annotateSpans(columns) // re-annotate the correct colSpans and rowSpans to render
    const rows = this._renderHeaderRowHelper(columns)
    if (select && rows.length) {
      rows[0].unshift(
        <th rowSpan={rows.length} key="all-selector" className="objectlist-table__th objectlist-table__th--border-bottom objectlist-table__th--selector">
          <AllSelector
            numSelected={numSelected}
            total={data.length}
            selectAll={this.handleSelectAll}
            deselectAll={this.handleDeselectAll}
          />
        </th>
      )
    }

    return (
      <thead className={rows.length > 1 ? 'objectlist-table__thead--grouped-headers' : ''}>
        {rows.map((row, index) => <tr key={`header-row-${index}`}>{row}</tr>)}
      </thead>
    )
  }

  _renderHeaderRowHelper = (columns) => {
    let currentRow = [...columns]
    let nextRow = []
    const rows = []
    let rowIndex = -1
    if (currentRow.length) {
      rows.push([])
      rowIndex++
    }
    let index = 0

    while (currentRow.length) {
      const header = currentRow.shift()
      if (header.hasOwnProperty('columns')) {
        nextRow = nextRow.concat(header.columns)
        rows[rowIndex].push(
          <th className={`objectlist-table__th`} key={`grouped-header-${header.fieldKey}`} colSpan={header._colSpan}>
            {header.header}
          </th>
        )
      } else {
        const label = Array.isArray(header) ? header[0].label : header.label
        const width = this.props.columnWidths[label] ? this.props.columnWidths[label].width : null
        rows[rowIndex].push(
          <TableHeader
            key={`header-${label}-${index}`}
            headerItems={header}
            label={label}
            width={width || header.width}
            rowSpan={header._rowSpan}
            sortKeys={this.props.meta.sortKeys}
            saveWidth={this.props.saveColumnWidth}
            updateSorting={this.props.updateSorting}
          />
        )
      }
      // move to the next row to render
      if (!currentRow.length) {
        currentRow = nextRow
        nextRow = []
        if (currentRow.length) {
          rows.push([])
          rowIndex++
        }
      }
      index++
    }
    return rows
  }

  _renderItemRowsHelper = (columns) => {
    const {data, selection, select, itemOnClick} = this.props
    const rowClasses = ['objectlist-table__row']
    if (itemOnClick) rowClasses.push('objectlist-table__row--clickable')
    return data.map((row, rowIndex) => {
      const selected = selection === ALL_SELECTED || row.id in selection
      // add row-specific classes defined in columns
      const rowClassesFromColumns = []
      const getRowClass = column => {
        if (Array.isArray(column)) { column.forEach(col => getRowClass(col)) }
        if (!column.rowClass) return
        if (typeof column.rowClass === 'function') { rowClassesFromColumns.push(column.rowClass(row)) }
        if (typeof column.rowClass === 'string') { rowClassesFromColumns.push(column.rowClass) }
      }
      columns.forEach(column => getRowClass(column))
      return (
        <tr
          key={`row-${rowIndex}`}
          className={ClassNames(rowClasses.concat(rowClassesFromColumns))}
          onClick={itemOnClick ? event => handleRowClick(event, row, itemOnClick) : null}
        >
          {select && (
            <td className="objectlist-table__td" key={`select-cell-${rowIndex}`} >
              <Selector
                toggleSelect={select}
                selected={selected}
                id={row.id}
              />
            </td>
          )}
          {columns.map((column, cellIndex) => {
            const toRender = Array.isArray(column) ? column : [column]
            const RenderedItems = []
            let i = 0
            while (i < toRender.length) {
              const RenderItem = toRender[i]
              if (RenderItem.item) {
                RenderedItems.push(RenderItem.item({
                  row: row,
                  column: RenderItem,
                  value: getValueFromAccessor(row, RenderItem.dataKey),
                  key: `item-${i}`,
                }))
              } else {
                RenderedItems.push(getValueFromAccessor(row, RenderItem.dataKey))
              }
              i++
            }
            return (
              <td key={`cell-${rowIndex}-${cellIndex}`} className="objectlist-table__td">
                {RenderedItems}
              </td>
            )
          })}
        </tr>
      )
    })
  }

  renderSummaryRow = (columns, key) => {
    const {summaryData, select} = this.props
    if (!summaryData) return null

    return (
      <tr
        key={`row-summary-${key}`}
        className="objectlist-table__row--summary"
      >
        {select && <td className="objectlist-table__td">-</td>}
        {columns.map((column, cellIndex) => {
          const toRender = Array.isArray(column) ? column : [column]
          const {summary, item} = toRender[0]
          let renderedItem = summary ? getValueFromAccessor(summaryData, summary) : null
          if (renderedItem && item) {
            renderedItem = item({
              row: summaryData,
              column: toRender[0],
              value: renderedItem,
              key: `summary-item`,
            })
          }
          return <td key={`cell-summary-${cellIndex}`} className="objectlist-table__td">{renderedItem || '-'}</td>
        })}
      </tr>
    )
  }

  handleSelectAll = () => {
    const {data, select, selection} = this.props
    select(data.map(row => row.id).filter(id => !(id in selection)))
  }

  handleDeselectAll = () => {
    this.props.select(null)
  }

  render() {
    const {status} = this.props
    const leafColumns = getLeafColumns(this.state.columns)
    return (
      <div className="objectlist-table--scroll">
        <Overlay status={status} />
        <table className="objectlist-table">
          {this.renderHeader()}
          <tbody>
            {this.renderSummaryRow(leafColumns, 0)}
            {this._renderItemRowsHelper(leafColumns)}
            {this.renderSummaryRow(leafColumns, 1)}
          </tbody>
        </table>
      </div>
    )
  }
}
