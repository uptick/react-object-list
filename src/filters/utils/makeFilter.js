import React from 'react'
import PropTypes from 'prop-types'
import FilterLabel from './FilterLabel'
import FilterComparison from './FilterComparison'
import RemoveFilter from './RemoveFilter'
import {FILTER_BASE_TYPE} from '../../utils/proptypes'

/**
 * Utility function to make a filter class with
 * the provided Renderer for its value input
 * @param  {React.Component} Renderer Component used to render input
 * @return {React.Component}          Component used to render filter
 */
const makeFilter = (Renderer) => {
  /**
   * Base filter class
   * Renderers common label, comparison and remove components
   */
  class Filter extends React.Component {
    static propTypes = {
      ...FILTER_BASE_TYPE,
      ...Renderer.propTypes,
      /** Function to be called when filter is updated */
      onChange: PropTypes.func.isRequired,
      /** Function to be called when filter is removed */
      removeFilter: PropTypes.func.isRequired,
      /** Options to be passed to the comparison dropdown */
      comparisonOptions: FilterComparison?.propTypes?.options,
      /** Currently selected comparison option */
      comparison: FilterComparison?.propTypes?.value,
      /** Icons */
      icons: PropTypes.object,
      /** Object of custom react-select styles */
      selectStyles: PropTypes.object,
    }

    static defaultProps = {
      comparisonOptions: [],
      icons: {},
      permanent: false,
      selectStyles: {},
    }

    /**
     * Handler to call onChange with the newly selected comparison
     * along with the other filter options - only if it has changed
     * @param newComparison Comparison just selected
     */
    onComparisonChange = (newComparison) => {
      const {filterKey, value, comparison} = this.props
      if (newComparison !== comparison) {
        this.props.onChange({filter: filterKey, comparison: newComparison, value: value})
      }
    }

    /**
     * Handler to call onChange with the new value
     * along with other filter options - only if it has changed
     * @param newValue Value just entered
     */
    onValueChange = (newValue) => {
      const {filterKey, value, comparison} = this.props
      if (newValue !== value) {
        this.props.onChange({filter: filterKey, comparison: comparison, value: newValue})
      }
    }

    /**
     * Handler to call removeFilter with the filter key
     * @param  {MouseEvent} event onClick event
     */
    removeFilter = (event) => {
      const {removeFilter, filterKey} = this.props
      event.preventDefault()
      removeFilter(filterKey)
    }

    render() {
      const {
        name, comparison, comparisonOptions, value, permanent, selectStyles, icons = {}, ...otherProps
      } = this.props
      return (
        <div className="objectlist-row">
          {name && <FilterLabel label={name} />}
          {!!comparisonOptions.length && (
            <FilterComparison
              options={comparisonOptions}
              value={comparison || comparisonOptions[0]}
              onChange={this.onComparisonChange}
              selectStyles={selectStyles}
            />
          )}
          <Renderer
            {...otherProps}
            comparison={comparison || comparisonOptions[0]}
            value={value}
            onChange={this.onValueChange}
            icons={icons}
            selectStyles={selectStyles}
          />
          {!permanent && (
            <RemoveFilter onClick={this.removeFilter} Icon={icons.RemoveFilter} />
          )}
        </div>
      )
    }
  }
  return Filter
}

export default makeFilter
