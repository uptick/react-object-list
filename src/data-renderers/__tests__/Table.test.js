import React from 'react'
import { shallow } from 'enzyme'
import { snapshotTest } from 'utils/tests'
import Table from '../Table'
import { getVisibleColumns, annotateSpans } from '../../utils/functions'

jest.mock('../TableHeader', () => 'TableHeader')
jest.mock('../Overlay', () => 'Overlay')
jest.mock('../../types/AllSelector', () => 'AllSelector')
jest.mock('../../types/Selector', () => 'Selector')

describe('Table', () => {
  const props = {
    select: jest.fn(),
    columns: [
      {
        dataKey: 'name',
        header: 'Name',
        item: ({row: {attributes: {name}}}) => name,
        optional: false,
        sortKey: 'üéÇ',
      },
      [{
        dataKey: 'age',
        header: 'Age',
        item: ({row: {attributes: {age}}, key}) => (<span key={key}>{age}</span>),
        optional: false,
        sortKey: 'ü•ß',
      },
      {
        dataKey: 'gender',
        header: 'Gender',
        item: ({row: {attributes: {gender}}}) => gender,
        optional: false,
        sortKey: 'üç©',
      }],
    ],
    columnWidths: {'age_gender': {width: 50}},
    saveColumnWidth: jest.fn(),
    data: [
      {id: 1, type: 'Person', attributes: {name: 'Sam', age: '25', gender: 'M'}},
      {id: 2, type: 'Person', attributes: {name: 'Mary', age: '12', gender: 'F'}},
      {id: 3, type: 'Person', attributes: {name: 'Q', age: '3', gender: '?'}},
      {id: 4, type: 'Person', attributes: {name: 'Peter', age: '111', gender: 'FM'}},
      {id: 5, type: 'Person', attributes: {name: 'Amy', age: '32', gender: 'Confused'}},
    ],
  }
  describe('Snapshots', () => {
    it('renders default', () => {
      snapshotTest(<Table {...props} select={null} />)
    })
    it('can select items', () => {
      snapshotTest(<Table {...props} />)
    })
    it('can click on items', () => {
      snapshotTest(<Table {...props} itemOnClick={jest.fn()} />)
    })
  })
  describe('Lifecycle', () => {
    it('componentWillReceiveProps', () => {
      const newProps = {
        columns: [
          {
            dataKey: 'firstname',
            header: 'First Name',
            item: ({row: {attributes: {name}}}) => name,
            optional: false,
            sortKey: 'üç™',
          },
          [{
            dataKey: 'years',
            header: 'Age in Years',
            item: ({row: {attributes: {age}}, key}) => (<span key={key}>{age}</span>),
            optional: false,
            sortable: 'üçÆ',
          }],
        ],
        meta: {
          extraColumns: ['firstname'],
        },
      }
      const instance = shallow(<Table {...props} />)
      instance.instance().setState({columns: []})
      instance.setProps(newProps)
      const visibleColumns = getVisibleColumns(newProps.columns, newProps.meta.extraColumns)
      annotateSpans(visibleColumns)
      expect(instance.instance().state.columns).toEqual(visibleColumns)
    })
  })
  describe('Functions', () => {
    it('handles select all', () => {
      spyOn(props, 'select')
      const instance = shallow(<Table {...props} />).instance()
      instance.handleSelectAll()
      expect(props.select).toHaveBeenCalledWith(props.data.map(row => row.id))
    })
    it('handles deselect all', () => {
      spyOn(props, 'select')
      const allSelection = {}
      props.data.forEach(row => { allSelection[row.id] = true })
      const instance = shallow(
        <Table
          {...props}
          numSelected={props.data.length}
          selection={allSelection}
        />
      ).instance()
      instance.handleDeselectAll()
      expect(props.select).toHaveBeenCalledWith(null)
    })
    it('handles select remaining', () => {
      spyOn(props, 'select')
      const mockSelection = {1: true, 3: true}
      const instance = shallow(
        <Table
          {...props}
          numSelected={2}
          selection={mockSelection}
        />
      ).instance()
      instance.handleSelectAll()
      expect(props.select).toHaveBeenCalledWith([2, 4, 5])
    })
  })
})
