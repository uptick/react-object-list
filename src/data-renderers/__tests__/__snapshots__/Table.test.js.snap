// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Snapshots can click on items 1`] = `
<div
  className="objectlist-table--scroll"
>
  <Overlay
    status="done"
  />
  <table
    className="objectlist-table"
  >
    <thead
      className=""
    >
      <tr>
        <th
          className="objectlist-table__th objectlist-table__th--border-bottom objectlist-table__th--selector"
          rowSpan={1}
        >
          <AllSelector
            deselectAll={[Function]}
            numSelected={undefined}
            selectAll={[Function]}
            total={5}
          />
        </th>
        <TableHeader
          headerItems={
            Object {
              "_colSpan": 1,
              "_rowSpan": 1,
              "dataKey": "name",
              "fieldKey": "name",
              "header": "Name",
              "item": [Function],
              "label": "name",
              "optional": false,
              "sortKey": "🎂",
            }
          }
          label="name"
          rowSpan={1}
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={undefined}
        />
        <TableHeader
          headerItems={
            Array [
              Object {
                "dataKey": "age",
                "fieldKey": "age_gender",
                "header": "Age",
                "item": [Function],
                "label": "age_gender",
                "optional": false,
                "sortKey": "🥧",
              },
              Object {
                "dataKey": "gender",
                "fieldKey": "gender_age",
                "header": "Gender",
                "item": [Function],
                "label": "gender_age",
                "optional": false,
                "sortKey": "🍩",
              },
            ]
          }
          label="age_gender"
          rowSpan={1}
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={50}
        />
      </tr>
    </thead>
    <tbody>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={1}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Sam
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            25
          </span>
          M
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={2}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Mary
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            12
          </span>
          F
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={3}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Q
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            3
          </span>
          ?
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={4}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Peter
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            111
          </span>
          FM
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={5}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Amy
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            32
          </span>
          Confused
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table Snapshots can select items 1`] = `
<div
  className="objectlist-table--scroll"
>
  <Overlay
    status="done"
  />
  <table
    className="objectlist-table"
  >
    <thead
      className=""
    >
      <tr>
        <th
          className="objectlist-table__th objectlist-table__th--border-bottom objectlist-table__th--selector"
          rowSpan={1}
        >
          <AllSelector
            deselectAll={[Function]}
            numSelected={undefined}
            selectAll={[Function]}
            total={5}
          />
        </th>
        <TableHeader
          headerItems={
            Object {
              "_colSpan": 1,
              "_rowSpan": 1,
              "dataKey": "name",
              "fieldKey": "name",
              "header": "Name",
              "item": [Function],
              "label": "name",
              "optional": false,
              "sortKey": "🎂",
            }
          }
          label="name"
          rowSpan={1}
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={undefined}
        />
        <TableHeader
          headerItems={
            Array [
              Object {
                "dataKey": "age",
                "fieldKey": "age_gender",
                "header": "Age",
                "item": [Function],
                "label": "age_gender",
                "optional": false,
                "sortKey": "🥧",
              },
              Object {
                "dataKey": "gender",
                "fieldKey": "gender_age",
                "header": "Gender",
                "item": [Function],
                "label": "gender_age",
                "optional": false,
                "sortKey": "🍩",
              },
            ]
          }
          label="age_gender"
          rowSpan={1}
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={50}
        />
      </tr>
    </thead>
    <tbody>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={1}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Sam
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            25
          </span>
          M
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={2}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Mary
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            12
          </span>
          F
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={3}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Q
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            3
          </span>
          ?
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={4}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Peter
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            111
          </span>
          FM
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          <Selector
            id={5}
            selected={false}
            toggleSelect={[MockFunction]}
          />
        </td>
        <td
          className="objectlist-table__td"
        >
          Amy
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            32
          </span>
          Confused
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table Snapshots renders default 1`] = `
<div
  className="objectlist-table--scroll"
>
  <Overlay
    status="done"
  />
  <table
    className="objectlist-table"
  >
    <thead
      className=""
    >
      <tr>
        <TableHeader
          headerItems={
            Object {
              "_colSpan": 1,
              "_rowSpan": 1,
              "dataKey": "name",
              "fieldKey": "name",
              "header": "Name",
              "item": [Function],
              "label": "name",
              "optional": false,
              "sortKey": "🎂",
            }
          }
          label="name"
          rowSpan={1}
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={undefined}
        />
        <TableHeader
          headerItems={
            Array [
              Object {
                "dataKey": "age",
                "fieldKey": "age_gender",
                "header": "Age",
                "item": [Function],
                "label": "age_gender",
                "optional": false,
                "sortKey": "🥧",
              },
              Object {
                "dataKey": "gender",
                "fieldKey": "gender_age",
                "header": "Gender",
                "item": [Function],
                "label": "gender_age",
                "optional": false,
                "sortKey": "🍩",
              },
            ]
          }
          label="age_gender"
          rowSpan={1}
          saveWidth={[MockFunction]}
          sortKeys={Array []}
          updateSorting={undefined}
          width={50}
        />
      </tr>
    </thead>
    <tbody>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          Sam
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            25
          </span>
          M
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          Mary
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            12
          </span>
          F
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          Q
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            3
          </span>
          ?
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          Peter
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            111
          </span>
          FM
        </td>
      </tr>
      <tr
        className="objectlist-table__row objectlist-table__row--clickable"
        onClick={[Function]}
      >
        <td
          className="objectlist-table__td"
        >
          Amy
        </td>
        <td
          className="objectlist-table__td"
        >
          <span>
            32
          </span>
          Confused
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
